### PR TITLE
Recruit - initial commit

### PR DESCRIPTION
**Pull request by Bibi40k**

### CODE REVIEW

File: recruit/organization.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
+package zohorecruit
```
Comment (schmorrison) [inline recruit/organization.go]: This package name does not seem to be in alignment with the other files in the package.

File: recruit/organization.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+import (
+	"fmt"
+	"time"
+
+	"git.ejobs.ro/core/mono/ejobs-zoho/zohoutils"
```
Comment (schmorrison) [inline recruit/organization.go]: I dont recognize this package, it seems to be used in quite a few places in the project.

**Review by schmorrison — CHANGES_REQUESTED** [review]
Thanks for the PR!!

Looks like you've done a lot of great work here. I noted a couple things in my first review of the changes. I'll look at everything a little more closely in the next couple days.

Thanks again and best regards,

Sam Morrison

File: recruit/organization.go (hunk 2 above)
Comment (Bibi40k) [reply recruit/organization.go]: It was from the project i have used, i have pushed another commit and i hope i fixed this, sorry.

**Comment by schmorrison** [comment]
Why closed?

**Comment by Bibi40k** [comment]
Who closed it ? It's my first pull request ever, if it's on me, it was 100% by mistake.

**Comment by schmorrison** [comment]
Reopened!

File: recruit/candidates.go [hunk 3]
```diff
@@ -0,0 +1,505 @@
...
+	}
+
+	return InsertCandidateResponse{}, fmt.Errorf("no 'InsertCandidateResponse' returned")
+}
+
+type InsertCandidateRequest struct {
+	Data    []ICRData `json:"data"`
+	Trigger []string  `json:"trigger"`
+}
+type ICRData struct {
```
Comment (schmorrison) [inline recruit/candidates.go:40]: I think, ideally, naming should follow the conventions.

File: recruit/candidates.go [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+
+import (
+	"encoding/json"
+
+	zoho "github.com/schmorrison/Zoho"
+)
+
+// https://www.zoho.eu/recruit/developer-guide/apiv2/insert-records.html
+func (c *API) InsertCandidates(request InsertCandidateRequest) (data InsertCandidateResponse, err error) {
+	endpoint := zohoutils.Endpoint{
```
Comment (schmorrison) [inline recruit/candidates.go:11]: I'm seeing `zohoutils` still about this package in several places.

Not really sure how it is being imported. Has it been defined as a global somewhere in the package?

File: recruit/candidates.go [hunk 5]
```diff
@@ -0,0 +1,505 @@
...
+		}
+	}
+
+	endpoint := zohoutils.Endpoint{
+		Name:         "candidates",
+		URL:          fmt.Sprintf("https://recruit.zoho.%s/recruit/v2/%s", c.ZohoTLD, CandidatesModule),
+		Method:       zohoutils.HTTPDelete,
+		ResponseData: &DeleteCandidateResponse{},
+		URLParameters: map[string]zohoutils.Parameter{
+			"ids": zohoutils.Parameter(idStr),
```
Comment (schmorrison) [inline recruit/candidates.go:375]: I do think having the constants and helper functions in a separate package is a good idea, and deserving of another PR (and probably major version).

File: recruit/records.go [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+// https://www.zoho.com/recruit/developer-guide/apiv2/upsert-records.html
+//
+// When performing an upsert, because the natural state of the records fields in this package is to 'omitempty' when encoding json,
+// if you want to empty the fields contents in zoho you will need to embed the records type in a struct in your own package,
+// and override the field with a field that has a json tag that does not contain 'omitempty'.
+// eg.
+//    type struct Candidate {
+//        zohorecruit.Candidate
+//        CustomField string `json:"Custom_Field"`
+//     }
```
Comment (schmorrison) [inline recruit/records.go:113]: Nice !!! 👍

Comment (schmorrison) [reply recruit/records.go:113]: I think the CRM package has this same API structure.

File: recruit/records.go [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+		NumberOfPositions        string        `json:"Number_of_Positions"`
+		Country                  interface{}   `json:"Country"`
+		CreatedBy                struct {
+			Name string `json:"name"`
+			ID   string `json:"id"`
+		} `json:"Created_By"`
+		IsAttachmentPresent bool `json:"Is_Attachment_Present"`
+	} `json:"data"`
+	Info PageInfo `json:"info,omitempty"`
+}
```
Comment (schmorrison) [inline recruit/records.go:232]: I may be misreading the API docs, but I believe this is another `interface{}` response type. The endpoint seems to allow "associating" any record with other records. The Job Openings & Candidates seems to be just an example.

Looking at the "Associate Candidate" endpoint, I could be completely wrong. What do you think?

File: recruit/records_types.go [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+	ConvertedDetail    struct {
+	} `json:"$converted_detail"`
+	CareerPageInviteStatus      string      `json:"Career_Page_Invite_Status"`
+	Mobile                      string      `json:"Mobile"`
+	LastName                    string      `json:"Last_Name"`
+	CurrentSalary               interface{} `json:"Current_Salary"`
+	AssociatedAnySocialProfiles bool        `json:"Associated_any_Social_Profiles"`
+	Fax                         string      `json:"Fax"`
+	ExpectedSalary              interface{} `json:"Expected_Salary"`
+}
```
Comment (schmorrison) [inline recruit/records_types.go:81]: Do these need to be `omitempty`? Will API return a fail if trying to `insert` an existing record?

File: recruit/recruit.go [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+	ClientsModule      Module = "Clients"
+	ContactsModule     Module = "Contacts"
+	CustomModule1      Module = "CustomModule1"
+	CustomModule2      Module = "CustomModule2"
+	CustomModule3      Module = "CustomModule3"
+	CustomModule4      Module = "CustomModule4"
+	CustomModule5      Module = "CustomModule5"
+	InterviewsModule   Module = "Interviews"
+	JobOpeningsModule  Module = "JobOpenings"  // this is used mostly
+	Job_OpeningsModule Module = "Job_Openings" // inconsistency in Zoho API
```
Comment (schmorrison) [inline recruit/recruit.go:22]: Ouch!

File: recruit/tags.go [hunk 10]
```diff
@@ -0,0 +1,379 @@
...
+func (c *API) CreateTags(request CreateTagsRequest, params map[string]zohoutils.Parameter) (data CreateTagsResponse, err error) {
+	endpoint := zohoutils.Endpoint{
+		Name:         "tags",
+		URL:          fmt.Sprintf("https://recruit.zoho.%s/recruit/v2/settings/tags", c.ZohoTLD),
+		Method:       zohoutils.HTTPPost,
+		ResponseData: &CreateTagsResponse{},
+		RequestBody:  request,
+		BodyFormat:   zohoutils.JSON,
+		URLParameters: map[string]zohoutils.Parameter{
+			"module": "", // mandatory
```
Comment (schmorrison) [inline recruit/tags.go:19]: You've clearly tested!

File: recruit/tags.go [hunk 11]
```diff
@@ -0,0 +1,379 @@
...
+		},
+	}
+
+	if len(params) > 0 {
+		for k, v := range params {
+			endpoint.URLParameters[k] = v
+		}
+	}
+
+	// pp.Printf("%s\n", endpoint)
```
Comment (schmorrison) [inline recruit/tags.go:29]: debug line, should remove

File: recruit/tags.go [hunk 12]
```diff
@@ -0,0 +1,379 @@
...
+		return *v, nil
+	}
+
+	return UpdateTagResponse{}, fmt.Errorf("data returned was not 'UpdateTagResponse'")
+}
+
+type UpdateTagRequest struct {
+	Tags []struct {
+		Name string `json:"name"`
+	} `json:"tags"`
```
Comment (schmorrison) [inline recruit/tags.go:279]: You defined a Tag struct at Line ```51```

File: recruit/user.go [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+		FirstName     interface{} `json:"first_name"`
+		Email         string      `json:"email"`
+		Status        string      `json:"status"`
+	} `json:"users"`
+	Info struct {
+		PerPage     int  `json:"per_page"`
+		Count       int  `json:"count"`
+		Page        int  `json:"page"`
+		MoreRecords bool `json:"more_records"`
+	} `json:"info"`
```
Comment (schmorrison) [inline recruit/user.go:100]: ```PageInfo``` struct?

File: recruit/jobopenings.go [hunk 14]
```diff
@@ -0,0 +1,456 @@
...
+
+	if v, ok := endpoint.ResponseData.(*JobOpeningsResponse); ok {
+		return *v, nil
+	}
+
+	return JobOpeningsResponse{}, fmt.Errorf("no 'JobOpeningsResponse' returned")
+}
+
+type JobOpeningsResponse struct {
+	Data []*JobOpening `json:"data,omitempty"`
```
Comment (schmorrison) [inline recruit/jobopenings.go:102]: Slice of pointer is not conventional. Is there a reason for that?

File: recruit/jobopenings.go [hunk 15]
```diff
@@ -0,0 +1,456 @@
...
+			}
+		}
+
+		res.Data = items
+
+		return *res, nil
+	}
+
+	return JobOpeningsResponse{}, fmt.Errorf("no 'JobOpeningsResponse' returned")
+}
```
Comment (schmorrison) [inline recruit/jobopenings.go:338]: That's unfortunate! You've clearly put alot of time and attention into this PR, I really appreciate it.

File: recruit/jobopenings.go [hunk 16]
```diff
@@ -0,0 +1,456 @@
...
+				Text string `xml:",chardata"`
+				No   string `xml:"no,attr"`
+				FL   []struct {
+					Text string `xml:",chardata"`
+					Val  string `xml:"val,attr"`
+				} `xml:"FL"`
+			} `xml:"row"`
+		} `xml:"JobOpenings"`
+	} `xml:"result"`
+}
```
Comment (schmorrison) [inline recruit/jobopenings.go:456]: I don't want to implement XML at the top level package, I was so happy when Zoho migrated to a REST-like API.

Does the recruit team plan to moved these endpoints to a JSON encoding?

**Review by schmorrison — CHANGES_REQUESTED** [review]
Lots of ```zohoutils```, a couple debug lines, but it all looks well thought out and faithful to the conventions that exist here.

The XML stuff is sad, but you seem to have figured it out. 

So just a couple things to iron out, I'll take another look, and it should be good to merge.

Thanks again, and best regards,

schmorrison

File: recruit/jobopenings.go (hunk 16 above)
Comment (Bibi40k) [reply recruit/jobopenings.go:456]: I had to use XML, it was the only way to get html formatting for job description. I hate it too :) Don't know about their intention. You're free to remove it from PR.

**Comment by Bibi40k** [comment]
I have made another mistake :( i have deleted [Bibi40k:recruit](https://github.com/Bibi40k/Zoho/tree/recruit) ... i didn't know where it came from. I thought i uploaded changes to my repo instead here, to yours.
How can i fix this, should i open another PR ?

**Comment by schmorrison** [comment]
> I have made another mistake :( i have deleted [Bibi40k:recruit](https://github.com/Bibi40k/Zoho/tree/recruit) ... i didn't know where it came from. I thought i uploaded changes to my repo instead here, to yours. How can i fix this, should i open another PR ?

Im honestly not sure what to do in that case. Presumably the fork/branch still exists somewhere cause the PR still exists. I wonder if you refork and push the code to that repo (same name as old repo), maybe Github is smart enough to figure it out. I imagine my review will be made "outdated" though.

You might have to do some Googleing on this one unfortunately.

**Comment by Bibi40k** [comment]
Can i delete PR and start it over ?
I already have fixed what you suggested

Bibi

**Comment by schmorrison** [comment]
Sure, push new changes to new PR. We will keep this opened for reference until all the comments are resolved.

**Comment by Bibi40k** [comment]
create new PR

**Pull request closed without merging**